### PR TITLE
DEV-256 Switch Easee integration to SingalR websocket

### PR DESCRIPTION
**Pull request by rstraczynski**

### CODE REVIEW

File: package/debian/DEBIAN/control [hunk 1]
```diff
@@ -1,5 +1,5 @@
 Package: easee
-Version: 0.0.9
+Version: 2.0.5
```
Comment (boguslaw-wojcik) [inline package/debian/DEBIAN/control:2]: Just a comment, it should be tagged as 2.1.0 as SIgnalR and other are new features.

File: package/debian/opt/thingsplex/easee/defaults/config.json [hunk 2]
```diff
@@ -17,5 +17,10 @@
   "pollingInterval": "30s",
   "slowChargingCurrentInAmperes": 10,
   "httpTimeout": "30s",
-  "observationsPeriod": "168h"
+  "signalR": {
+    "connCreationTimeout": "5s",
```
Comment (boguslaw-wojcik) [inline package/debian/opt/thingsplex/easee/defaults/config.json:21]: Where did you get these numbers?

File: src/internal/easee/connector.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+func (c *connector) Disconnect(_ adapter.Thing) {
+	c.manager.Unregister(c.chargerID)
+}
+
+func (c *connector) Connectivity() *adapter.ConnectivityDetails {
+	if c.signalRClient.Connected() {
+		return &adapter.ConnectivityDetails{
+			ConnectionStatus: adapter.ConnectionStatusUp,
```
Comment (boguslaw-wojcik) [inline src/internal/easee/connector.go:52]: Please add here ConnectionType: Indirect

File: src/internal/easee/connector.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+		}
+	}
+
+	return &adapter.ConnectivityDetails{
+		ConnectionStatus: adapter.ConnectionStatusDown,
```
Comment (boguslaw-wojcik) [inline src/internal/easee/connector.go:57]: The same as above.

File: src/internal/easee/connector.go [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+func (c *connector) extractMeterElecServices(thing adapter.Thing) ([]meterelec.Service, error) {
+	raw := thing.Services(meterelec.MeterElec)
+	meterElecs := make([]meterelec.Service, 0, len(raw))
+
+	for _, service := range raw {
+		me, ok := service.(meterelec.Service)
+		if !ok {
+			return nil, fmt.Errorf("expected a service to be a meter_elec, got %s instead", reflect.TypeOf(service))
```
Comment (boguslaw-wojcik) [inline src/internal/easee/connector.go:153]: Tip: You can also always use %T which is the same.

File: src/internal/easee/controller.go [hunk 6]
```diff
@@ -96,163 +77,39 @@ func (c *controller) SetChargepointCableLock(locked bool) error {
...
-	if !ok {
-		return 0, errors.Errorf("expected float64, got %s instead", reflect.TypeOf(rawEnergy))
-	}
-
-	return energy, nil
-}
-
 // backoff allows Easee cloud to process the request and invalidates local cache.
 func (c *controller) backoff() {
 	time.Sleep(c.cfgService.GetEaseeBackoff())
```
Comment (boguslaw-wojcik) [inline src/internal/easee/controller.go:114]: Do we even need backoff now?

File: src/internal/easee/signalr.go [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+	return &signalRManager{
+		client:   client,
+		receiver: receiver,
+		chargers: make(map[string]chargerItem),
+		done:     make(chan struct{}),
+	}
+}
+
+func (m *signalRManager) Start() error {
+	go m.run()
```
Comment (boguslaw-wojcik) [inline src/internal/easee/signalr.go:51]: This is a nitpick and should not make any difference as well as is not per se requirement here from me, but:
Start/stop services should be protected from being started/stopped multiple times. Executing start multiple times will have funny consequences here.

The most proper solution is to either have an idempotent start/stop or erroring when being used in an incompatible phase, depending on the use case.

File: src/internal/easee/signalr.go [hunk 8]
```diff
@@ -0,0 +1,451 @@
...
+					item.cache.setConnected(false)
+				}
+
+				m.mu.RUnlock()
+
+				continue
+			}
+
+			for chargerID, charger := range m.chargers {
+				charger.cache.setConnected(true)
```
Comment (boguslaw-wojcik) [inline src/internal/easee/signalr.go:126]: Why this is before the error? Is there any reason for it?

File: src/internal/easee/signalr.go [hunk 9]
```diff
@@ -0,0 +1,451 @@
...
+
+	cb()
+}
+
+type chargerItem struct {
+	cache     ObservationCache
+	callbacks map[ObservationID]func()
+}
+
+// TODO: move it to a signalR client.
```
Comment (boguslaw-wojcik) [inline src/internal/easee/signalr.go:243]: Todo?

**Review by boguslaw-wojcik — CHANGES_REQUESTED** [review]
Small changes and some questions.

File: package/debian/opt/thingsplex/easee/defaults/config.json (hunk 2 above)
Comment (rstraczynski) [reply package/debian/opt/thingsplex/easee/defaults/config.json:21]: Apart from `connCreationTimeout` and `invokeTimeout`, these are the defaults suggested by `philippseith/signalr`. Do you suggest to adapt it?

File: src/internal/easee/connector.go (hunk 3 above)
Comment (rstraczynski) [reply src/internal/easee/connector.go:52]: Done

File: src/internal/easee/connector.go (hunk 4 above)
Comment (rstraczynski) [reply src/internal/easee/connector.go:58]: Done

File: src/internal/easee/controller.go (hunk 6 above)
Comment (rstraczynski) [reply src/internal/easee/controller.go:114]: Nope, removed

File: src/internal/easee/connector.go [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+		manager:       manager,
+		httpClient:    httpClient,
+		signalRClient: signalRClient,
+		chargerID:     chargerID,
+		cache:         cache,
+	}
+}
+
+func (c *connector) Connect(t adapter.Thing) {
+	callbacks, err := c.signalRCallbacks(t)
```
[the remark this answers is not in the document]
Comment (rstraczynski) [reply src/internal/easee/connector.go:35]: In the next release :D

File: src/internal/easee/signalr.go (hunk 7 above)
Comment (rstraczynski) [reply src/internal/easee/signalr.go:52]: Start/Stop should be called only once in this case, but I'll make it idempotent to be safe. Thanks for the hint!

File: src/internal/easee/signalr.go (hunk 8 above)
Comment (rstraczynski) [reply src/internal/easee/signalr.go:150]: We discussed that on PM - there's no ideal solution here.

**Review by boguslaw-wojcik — COMMENTED** [review]
I don't see tasks for adapter anywhere.

**Review by boguslaw-wojcik — APPROVED** [review]

**Pull request merged**